lib/galaxy/jobs/runners/drmaauniva.py

```diff
@@ -0,0 +1,640 @@
+            mem_wasted = extinfo["memory_wasted"]
+
+            # check if the output contains indicators for a memory violation
+            memviolation = _check_memory_limit( ajs.error_file )
+            logging.debug("DRMAAUnivaJobRunner:_complete_terminal_job ({jobid}) memviolation {mv}".format(jobid=ajs.job_id, mv=memviolation))
+
+            # check job for run time or memory violation
+            if "deleted" in extinfo and extinfo[ "deleted" ]:
+                log.info( '(%s/%s) Job was cancelled (e.g. with qdel)', ajs.job_wrapper.get_id_tag(), ajs.job_id )
+                ajs.fail_message = "This job failed because it was cancelled by an administrator."
+                drmaa_state = self.drmaa.JobState.FAILED
+            elif ("signal" in extinfo and extinfo["signal"] == "SIGKILL") and time_wasted > time_granted:
+                log.info( '({tag}/{jobid}) Job hit walltime'.format(tag=ajs.job_wrapper.get_id_tag(), jobid=ajs.job_id) )
+                ajs.fail_message = "This job was terminated because it ran longer than the maximum allowed job run time."
+                ajs.runner_state = ajs.runner_states.WALLTIME_REACHED
+                drmaa_state = self.drmaa.JobState.FAILED
+            elif memviolation or mem_wasted > mem_granted:
+                log.info( '({idtag}/{jobid}) Job hit memory limit ({used}>{limit})'.format(idtag=ajs.job_wrapper.get_id_tag(), jobid=ajs.job_id, used=mem_wasted, limit=mem_granted) )
+                ajs.fail_message = "This job was terminated because it used more than the maximum allowed memory."
+                ajs.runner_state = ajs.runner_states.MEMORY_LIMIT_REACHED
+                drmaa_state = self.drmaa_job_states.FAILED
+        elif state in [ self.drmaa.JobState.QUEUED_ACTIVE, self.drmaa.JobState.SYSTEM_ON_HOLD, self.drmaa.JobState.USER_ON_HOLD, self.drmaa.JobState.USER_SYSTEM_ON_HOLD, self.drmaa.JobState.RUNNING, self.drmaa.JobState.SYSTEM_SUSPENDED, self.drmaa.JobState.USER_SUSPENDED ]:
+            log.warning( '({tag}/{jobid}) Job is {state}, returning to monitor queue'.format(tag=ajs.job_wrapper.get_id_tag(), jobid=ajs.job_id, state=self.drmaa_job_state_strings[state] ))
+            # TODO return True?
+            return True  # job was not actually terminal
+        else:
+            logging.error( "DRMAAUniva: job {job_id} determined unknown state {state}".format(job_id=ajs.job_id, state=state) )
+            drmaa_state = self.drmaa_job_states.FAILED
+        # by default, finish the job with the state from drmaa
+        return super( DRMAAUnivaJobRunner, self )._complete_terminal_job( ajs, drmaa_state=drmaa_state )
+
+    def _drmaa_state_is_refined( self, statep, staten ):
+        """
+        check if staten is more 'severe' than statep
+        """
+        # definition of the severity of job states, the hex codes are
+        # from the drmaa C library which seem to define a useful order
+        drmaa_job_state_order = {
```

Comment:
Can this be merged with `drmaa_job_state_strings` in the base DRMAA runner? Perhaps the values should be another dict with keys `code` and `message`? Or if you wanted to be really fancy, make it a class and define the comparison operator methods so you can then just compare instances of the class.

Reply:
Correct, the cleanest solution would be a class. But then I would need to transform all job states produced by any DRMAARunner into JobState objects .. seems quite a bit of work and I can only test the standard DRMAA runner and the Univa runner. 

Is it worth the effort? Or would you rather prefer the dict based approach that you have suggested?